### DISCUSSION TITLE
Making CapRover resilient to captain node failure

### DISCUSSION THREAD

**Discussion by khromov** [caprover/caprover#1524] [Ideas]
As mentioned in https://github.com/caprover/caprover/discussions/1029 , when you have configured a swarm cluster in CapRover then the captain node is the single point of failure.

Could one possible way of dealing with this be to add a third party Load Balancer such as the AWS ALB or DigitalOcean Load Balancer. These managed services can distribute traffic amongst several servers running Docker/CapRover and can handle SSL termination. 

But there would have to be some way for each node to expose a port with the running application. Could this perhaps be done already using Port Mapping? Do you see any pitfalls to this approach? To run this in production it would be important to be able to limit which IP:s can contact the ports, so that only your load balancer can connect with them.

Another solution (that would be even simpler) is if each node in the swarm could have a copy of the SSL certificate/Nginx config and terminate connections. In such no load balancer is needed, you would simply add all your node IP:s to the DNS record for your domain and each visiting browser [would act as a rudimentary load balancer, trying multiple IP:s to reach your service](https://serverfault.com/questions/247885/browser-based-dns-failover-using-multiple-a-records). 

Could any of these two options be viable for CapRover?

**Comment by githubsaturn** [comment]
In theory, doable, but in practice, CapRover isn't designed for this. You'll end up having to do so much manual work to a degree that I'd argue CapRover just makes the process slower than providing any help.

- You'll have to run some sort of rsync or nfs to make sure `/captain` directory is in sync.
- You'll have to update the `captain-*` services so that they are not locked to a nodeID, rather, they are locked to role=manager

**Comment by khromov** [comment]
@githubsaturn Thanks for the response, I have tried an approach using an external load balancer and it was very easy to set up with CapRover, so I thought I'd share:

### Basic assumptions

* We want to protect an application from single point of failure, for example we might want to provision 3 servers in different data centers and for the application to continue being available if 1-2 servers go down.
* We do not have to run services with persistent volumes.
* We can use a third party ingress solution that manages SSL certificates.

### Setup

For this test I used DigitalOcean with their managed [Load Balancer](https://www.digitalocean.com/products/load-balancer) service, but this approach should work for Cloudflare, AWS Cloudfront or any similar services.

First, we set up a Captain node and two child node and connect them in a cluster. 

We then set up an application on CapRover. We configure a port mapping (in this case 8888). This will make each node IP on port 8888 route to the application. I am not an expert on Swarm but it seems to use the [routing mesh](https://docs.docker.com/engine/swarm/ingress/) because when connecting to a node you will actually get a response from a random node in the swarm.

![2022-10-01 02_13_33-CapRover _ Server Dashboard — Mozilla Firefox](https://user-images.githubusercontent.com/1207507/193373809-64372b15-d295-4ac4-a29f-7a6d544af200.png)

Now we can create a Load Balancer in DigitalOcean. We add our captain + worker nodes to the load balancer on port 8888. 

![2022-10-01 01_37_48-Control Panel - DigitalOcean — Mozilla Firefox](https://user-images.githubusercontent.com/1207507/193374785-87a2a30c-bece-44b4-b162-8543950d5510.png)

We verify that the nodes are healthy. We can now configure SSL termination for the Load Balancer.

![2022-10-01 01_54_01-Control Panel - DigitalOcean — Mozilla Firefox](https://user-images.githubusercontent.com/1207507/193374858-e5574149-2129-4387-94d4-ee6247fa9353.png)

I have tried disconnecting both worker nodes and the manager node, and it works just fine (broken nodes are taken out of the load balancer when the health checks start to fail, and re-join when they return back from being offline).

**Reply by githubsaturn** [reply]
This is a great step! Congrats!

The only downside with this approach is that if your main node goes down, you can't connect to CapRover again. If that's not a concern for you, your approach works great!

**Reply by khromov** [reply]
@githubsaturn You are correct, the scope of this redundant setup is a temporary drop-out of one or more nodes. (Connectivity problems,  power outage, etc.)

You will still need to have backups of the Captain node in case of a catastrophic failure where it dies and you can't boot it again, and you won't really be able to manage your app if the captain node drops out. 😅 But it's good enough for my use case!
